bug-fix/continue-watching-translation

ahmedNaser7 opened the pull request:
fix the issue of language in continue watching list to make it change from arabic to english base on use selection 
handle the caching data 

Refactor: Improve `getContinueWatchingMedia` to use `MovieRepository` and `TvShowRepository`

This commit refactors the `getContinueWatchingMedia` function in `MediaRepositoryImpl` to utilize `MovieRepository` and `TvShowRepository` for fetching detailed movie and TV show information respectively.

The following changes were made:
- Injected `MovieRepository` and `TvShowRepository` into `MediaRepositoryImpl`.
- Updated the Hilt module to provide these dependencies.
- Modified `getContinueWatchingMedia` to:
    - Separate media items into movies and TV shows.
    - Fetch detailed information for each movie and TV show using their respective repositories.
    - Combine the results back into a list of `Media` objects.
- Added `toMedia()` extension functions to `Movie` and `TvShow` domain entities in `MediaMapper.kt` for easier conversion.



https://github.com/user-attachments/assets/5baa3581-6c31-46b7-9c53-7a2c7fd145ea

ahmedNaser7 committed:
Refactor: Improve `getContinueWatchingMedia` to use `MovieRepository` and `TvShowRepository`

This commit refactors the `getContinueWatchingMedia` function in `MediaRepositoryImpl` to utilize `MovieRepository` and `TvShowRepository` for fetching detailed movie and TV show information respectively.

The following changes were made:
- Injected `MovieRepository` and `TvShowRepository` into `MediaRepositoryImpl`.
- Updated the Hilt module to provide these dependencies.
- Modified `getContinueWatchingMedia` to:
    - Separate media items into movies and TV shows.
    - Fetch detailed information for each movie and TV show using their respective repositories.
    - Combine the results back into a list of `Media` objects.
- Added `toMedia()` extension functions to `Movie` and `TvShow` domain entities in `MediaMapper.kt` for easier conversion.

ahmedNaser7 committed:
update release note

ahmedNaser7 committed:
Refactor: Remove unused MediaEntity.toDomain() extension function

The `MediaEntity.toDomain()` extension function in `MediaMapper.kt` was not being used and has been removed to improve code clarity and reduce unused code.

ahmedNaser7 committed:
Test: Inject MovieRepository and TvShowRepository in MediaRepositoryImplTest

The `movieRepository` and `tvShowRepository` dependencies were added to the `MediaRepositoryImplTest` to align with changes in the `MediaRepositoryImpl` constructor.

Joseph-Sameh-0 commented on repository/media/src/main/java/com/repository/media/repository/MediaRepositoryImpl.kt:
```diff
@@ -116,13 +124,24 @@ class MediaRepositoryImpl(
-        }
-            .catch {
-                throw FailedException("getContinueWatchingMedia")
+    override  fun getContinueWatchingMedia(): Flow<List<Media>> {
+
+        return mediaLocalDataSource.getMediaContinueWatching().map { mediaList ->
+
+            val movies = mediaList.filter { it.type == MediaTypeEntity.Movie }
+            val tvShows = mediaList.filter { it.type == MediaTypeEntity.TvShow }
+
+            val filteredMovies = CoroutineScope(IO).async {
+                movies.map { movieRepository.getMovieDetails(it.id) }
             }
+
+            val filteredTvShows = CoroutineScope(IO).async {
+                tvShows.map { tvShowRepository.getTvShowDetails(it.id) }
+            }
+
+            filteredMovies.await().map { it.toMedia() } + filteredTvShows.await().map { it.toMedia() }
+        }
+
```
It's not good to do this heavy operation 
it will slowdown the response
you can save by the two languages when go to movie or tv show and get what you want

ahmedNaser7 replied:
Nah its not heavy its all about the list of continue watching , all you do is filteration and getDetails , try to add a big list and test if the response take a lot of time or not

Joseph-Sameh-0 replied:
it will take a lot of time after 1 hour when the cached data deleted

ahmedNaser7 replied:
@Joseph-Sameh-0 
you can delete cache and try use it

ahmedNaser7 committed:
update media repo test

ahmedNaser7 committed:
Refactor: Optimize continue watching media retrieval

This commit refactors the `getContinueWatchingMedia` function in `MediaRepositoryImpl.kt` to improve efficiency and error handling.

Key changes:
- Retrieves language setting asynchronously.
- Checks for existing local media for the "CONTINUE_WATCHING" category before fetching details from remote sources.
- Uses `awaitAll` to concurrently fetch movie and TV show details when local data is not available.
- Persists fetched continue watching media to the local database.
- Adds error handling with a `catch` block to throw a `FailedException`.
- Adds `CONTINUE_WATCHING` to the `Category` enum in `HomeMediaEntity.kt`.

ahmedNaser7 committed:
Merge branch 'develop' into bug-fix/continue-watching-translation

# Conflicts:
#	repository/media/src/main/java/com/repository/media/mapper/MediaMapper.kt
#	repository/media/src/main/java/com/repository/media/repository/MediaRepositoryImpl.kt
#	repository/media/src/test/java/com/repository/media/repository/MediaRepositoryImplTest.kt

ahmedNaser7 committed:
Merge branch 'develop' into bug-fix/continue-watching-translation

# Conflicts:
#	app/release-info.txt

ahmedNaser7 committed:
Fix package names for movie and tvshow

The commit addresses incorrect package names for `MovieRepository`, `TvShowRepository`, `Movie`, and `TvShow`. These were previously using `com.paris_2.domain.media` and have been corrected to `com.paris.domain.media`.

ahmedNaser7 committed:
Merge branch 'develop' into bug-fix/continue-watching-translation

# Conflicts:
#	repository/media/src/main/java/com/repository/media/repository/MediaRepositoryImpl.kt

ahmedNaser7 committed:
Remove duplicate Category import in MediaRepositoryImpl.kt

ahmedNaser7 committed:
Fix: Corrected import paths in MediaRepositoryImplTest.kt

The import paths for `MovieRepository` and `TvShowRepository` were updated from `com.paris_2.domain.media.repository` to `com.paris.domain.media.repository`.
